src/Stratis.Bitcoin.Features.Miner/PowMining.cs

```diff
@@ -168,110 +168,181 @@ public void StopMining()
         }
 
         ///<inheritdoc/>
-        public List<uint256> GenerateBlocks(ReserveScript reserveScript, ulong generate, ulong maxTries)
+        public List<uint256> GenerateBlocks(ReserveScript reserveScript, ulong amountOfBlocksToMine, ulong maxTries)
         {
-            ulong nHeightStart = 0;
-            ulong nHeightEnd = 0;
-            ulong nHeight = 0;
+            var context = new MineBlockContext(amountOfBlocksToMine, (ulong)this.chain.Height, maxTries, reserveScript);
 
-            nHeightStart = (ulong)this.chain.Height;
-            nHeight = nHeightStart;
-            nHeightEnd = nHeightStart + generate;
-            int nExtraNonce = 0;
-            var blocks = new List<uint256>();
-
-            while (nHeight < nHeightEnd)
+            while (context.MiningCanContinue)
             {
-                this.miningCancellationTokenSource.Token.ThrowIfCancellationRequested();
-
-                ChainedHeader chainTip = this.consensusLoop.Tip;
-                if (this.chain.Tip != chainTip)
-                {
-                    Task.Delay(TimeSpan.FromMinutes(1), this.nodeLifetime.ApplicationStopping).GetAwaiter().GetResult();
+                if (!ConsensusIsAtTip(context))
                     continue;
-                }
-
-                BlockTemplate blockTemplate = this.blockProvider.BuildPowBlock(chainTip, reserveScript.ReserveFullNodeScript);
 
-                if (this.network.Consensus.IsProofOfStake)
-                {
-                    // Make sure the POS consensus rules are valid. This is required for generation of blocks inside tests,
-                    // where it is possible to generate multiple blocks within one second.
-                    if (blockTemplate.Block.Header.Time <= chainTip.Header.Time)
-                        continue;
-                }
+                if (!BuildBlock(context))
+                    continue;
 
-                nExtraNonce = this.IncrementExtraNonce(blockTemplate.Block, chainTip, nExtraNonce);
-                Block block = blockTemplate.Block;
+                if (!MineBlock(context))
```

Comment:
Reminder to add `this.` on all 3 mtheods

Reply:
👍